pkg/certificates/authorizations/db.go

```diff
@@ -0,0 +1,271 @@
+		var listItem storage.ListItem
+		for iterator.Next(ctx, &listItem) {
+			userIDs = append(userIDs, listItem.Key.String())
+		}
+		return nil
+	})
+	return userIDs, err
+}
+
+// List returns all authorizations in the database.
+func (authDB *DB) List(ctx context.Context) (auths Authorizations, err error) {
+	defer mon.Task()(&ctx)(&err)
+	err = authDB.db.Iterate(ctx, storage.IterateOptions{
+		Recurse: true,
+	}, func(ctx context.Context, iterator storage.Iterator) error {
+		var listErrs errs.Group
+		var listItem storage.ListItem
+		for iterator.Next(ctx, &listItem) {
+			var nextAuths Authorizations
+			if err := nextAuths.Unmarshal(listItem.Value); err != nil {
+				listErrs.Add(err)
+			}
+			auths = append(auths, nextAuths...)
+		}
+		return listErrs.Err()
+	})
+	return auths, err
+}
+
+// Claim marks an authorization as claimed and records claim information.
+func (authDB *DB) Claim(ctx context.Context, opts *ClaimOpts) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	now := time.Now().Unix()
+	if !(now-MaxClaimDelaySeconds < opts.Req.Timestamp) ||
+		!(opts.Req.Timestamp < now+MaxClaimDelaySeconds) {
+		return ErrAuthorization.New("claim timestamp is outside of max delay window: %d", opts.Req.Timestamp)
+	}
+
+	ident, err := identity.PeerIdentityFromPeer(opts.Peer)
+	if err != nil {
+		return err
+	}
+
+	peerDifficulty, err := ident.ID.Difficulty()
+	if err != nil {
+		return err
+	}
+
+	if peerDifficulty < opts.MinDifficulty {
+		return ErrAuthorization.New("difficulty must be greater than: %d", opts.MinDifficulty)
+	}
+
+	token, err := ParseToken(opts.Req.AuthToken)
+	if err != nil {
+		return err
+	}
+
+	auths, err := authDB.Get(ctx, token.UserID)
+	if err != nil {
+		return err
+	}
+
+	for i, auth := range auths {
+		if auth.Token.Equal(token) {
+			if auth.Claim != nil {
+				return ErrAuthorization.New("authorization has already been claimed: %s", auth.String())
+			}
+
+			auths[i] = &Authorization{
+				Token: auth.Token,
+				Claim: &Claim{
+					Timestamp:        now,
+					Addr:             opts.Peer.Addr.String(),
+					Identity:         ident,
+					SignedChainBytes: opts.ChainBytes,
+				},
+			}
+			if err := authDB.put(ctx, token.UserID, auths); err != nil {
+				return err
+			}
+			break
+		}
+	}
+	return nil
+}
+
+// Unclaim removes a claim from an authorization.
+func (authDB *DB) Unclaim(ctx context.Context, authToken string) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	token, err := ParseToken(authToken)
+	if err != nil {
+		return err
+	}
+
+	auths, err := authDB.Get(ctx, token.UserID)
+	if err != nil {
+		return err
+	}
+
+	for i, auth := range auths {
+		if auth.Token.Equal(token) {
+			auths[i].Claim = nil
+			return authDB.put(ctx, token.UserID, auths)
+		}
+	}
+	return errs.New("token not found in authorizations DB")
+}
+
+func (authDB *DB) add(ctx context.Context, userID string, newAuths Authorizations) (err error) {
+	defer mon.Task()(&ctx)(&err)
```

Comment:
```suggestion
	defer mon.Task()(&ctx, userID)(&err)
```

Reply:
I think we may not want to call `mon.Task` on these unexported methods. What do you think?

Reply:
Not sure, but I don't think that hurt and the last conversation that I had with @jtolds regarding `mon.Task` calls,  I understood that's good to call almost everywhere.

Reply:
ok